### ISSUE TITLE
Equal floats and ints have different hashes

### ISSUE THREAD

**Issue by oprypin** [crystal-lang/crystal#2818]
First, the immediate problem caused by this fact:

``` crystal
{0 => :a, 0.0 => :b, 1 => :c, 1.0 => :d}
# => {0 => :b, 1 => :c, 1.0 => :d}
```

Because:

``` crystal
0 == 0.0  &&  1 == 1.0
# => true
{0.hash, 0.0.hash}
# => {0, 0}
{1.hash, 1.0.hash}
# => {1, 4607182418800017408}
```

(there don't seem to be any other such "hash collisions" other than 0 but my search wasn't exhaustive, and this is still dangerous).

In case this wasn't clear, the problem is the definition of hashes is violated &ndash; equal values must have equal hashes.

I see a few general approaches to solving this:
1. Like Ruby: https://github.com/crystal-lang/crystal/pull/2809#issuecomment-225435084  
   Introduce a "hash equality" concept and explicitly mark Ints and Floats to be not equal in the hashing context.  
   (The example above would have 4 pairs in the end)
2. Variation on the above:
   Make different types never be equal in the hashing context. How to determine whether types are "same" or "different"? Again, a few suggestions:
   1. It's the exact same type
   2. The types have the same hash method (same type or inherited).
3. Like Python: https://github.com/crystal-lang/crystal/pull/2809#issuecomment-225435854  
   Set up the hashing functions in a way that Ints and Floats have the same hash.  
   (The example above would have 2 pairs in the end)  
   Theoretically we could take [their algorithm](https://github.com/python/cpython/blob/2d264235/Python/pyhash.c#L86) but not sure about licensing.  
   To elaborate what Python does:

``` python
>>> hash(234), hash(234.0), hash(Decimal('234.0'))
(234, 234, 234)
>>> hash(0.5), hash(Fraction(1, 2)), hash(Decimal('0.5'))
(1152921504606846976, 1152921504606846976, 1152921504606846976)
>>> {234: 1, 234.0: 2, Decimal('234.0'): 3}
{234: 3}
```

**Comment by david50407** [comment]
+1 for ruby-like solution. But it's fine now, C# and Java also make the hash of `0` and `0.0` the same. 

IMO, Crystal gets everything to be objects, so the hash should be different when these two items having different type even their data is the same in the memory at least, so `0`'s hash will never equals to `0.0`'s, cause their types are not the same at all.

**Comment by ozra** [comment]
Without diving in to the problem, at a glance I'd say it's not a problem with hashing (duplicates are expected - it's not meant to be perfect hashing, just fast and distribute over buckets and similar uses).
The problem with the hash-map must be because `0 == 0.0` - so this instead puts some more fuel on https://github.com/crystal-lang/crystal/issues/567#issuecomment-223762748 (don't allow float <=> int promotions)

**Comment by akzhan** [comment]
I think that it should be closed because #5276 has been merged.
